README.md
[![Build Status](https://travis-ci.com/evenset/safe-auth.svg?branch=development)](https://travis-ci.com/evenset/safe-auth)
[![codecov](https://codecov.io/gh/evenset/safe-auth/branch/development/graph/badge.svg)](https://codecov.io/gh/evenset/safe-auth)
[![npm version](https://badge.fury.io/js/safe-auth.svg)](https://badge.fury.io/js/safe-auth)

# What's safe-auth

safe-auth is a general purpose authentication solution for Node.js.
It's written in TypeScript with %100 test code coverage.

# Features

At its core it provides these features:

1. Create users
1. Activate/Deactivate users
1. Authenticating a user based on their username and password and returning
    the authenticated user
1. Issuing a token for a user
1. Authenticating a user based on an active issued token and returning the
    authenticated user
1. Issuing new token based on a refresh token (signing in)
1. Reset user's password when they forget it
1. Revoking tokens (signing out)
1. Revoking all active tokens of a user (signing out of all devices)
1. Optionally block a user when he has several failed attempts
1. Logging all requests, responses and events
1. Configuring paramters and enabaling/disabling optional features in a
    centralized config system

# Storage engines

By default safe-auth provides an in memory storage engine (not to be used in
production) and it's able to use different storage engines via third party
packages.

Currently there is a third party package that uses Sequelize as the storage
backend which makes Postgres, MySQL, MariaDB, SQLite and Microsoft SQL
available.

TypeORM and Mongoose will be supported by other third party packages soon. So
the goal is to have at least these storage backends for safe-auth:
1. Sequelize by [safe-auth-sequelize](https://github.com/evenset/safe-auth-sequelize)
2. TypeORM - to be done in future
3. Mongoose - to be done in future

# Serving options

The core has its API with the JS classes it exposes. One can simply use these
classes in her project. But there are third party packages that provide
different serving options ready out of the box:
1. HTTP POST/GET requests
1. GraphQL
1. Rest

# Reporting bugs

You can report issues/bugs in the github repository of the project:
https://github.com/evenset/safe-auth/issues


CHANGELOG.md
---

## 0

### 0.2.0    2019-05-11

- Added CHANGELOG.md
- Made `getAccessToken`, `getAccessTokens` and `getActiveAccessTokens`
    abstract in `User` model so that each storage implementation implements
    them the way it makes more sense (needed to implement safe-auth-sequelize)
- Implemented `getAccessToken`, `getAccessTokens` and `getActiveAccessTokens`
    in `MemoryUser`


---


src/tests/integrity.test.ts
import chai from 'chai';
import chaiAsPromised from 'chai-as-promised';
import sinon from 'sinon';
import sinonChai from 'sinon-chai';

import {MemoryAccessToken, MemoryUser} from '../memory-storage/index';

chai.use(chaiAsPromised);
chai.use(sinonChai);
const expect = chai.expect;

afterEach((): void => {
    sinon.restore();
    MemoryUser.items = {};
    MemoryAccessToken.items = {};
});

describe('Normal flow', (): void => {
    it('should be able to create a user, authenticate her, issue an access' +
        ' token for her and authenticate the access token, each of these' +
        ' steps should return null when the previous steps aren\'t done yet' +
        ' or when called with invalid arguments', async (): Promise<void> => {
        expect(await MemoryUser.authenticate('u', 'p'))
            .to.be.null;
        const user = new MemoryUser({username: 'u', password: 'p'});
        expect(await MemoryUser.authenticate('u', 'p'))
            .to.be.null;
        await user.save();
        expect(await MemoryUser.authenticate('u', 'p'))
            .to.be.equal(null);
        user.isActive = true;
        // Event though saving is not required in memory storage backend it's
        // here to keep the code snippet a valid sample for all storage backends
        await user.save();
        expect(await MemoryUser.authenticate('u', 'p'))
            .to.be.equal(user);

        expect(await MemoryAccessToken.authenticate('some-token'))
            .to.be.null;
        const accessToken = await MemoryAccessToken.issue(user);
        expect(await MemoryAccessToken.authenticate('some-token'))
            .to.be.null;
        expect(await MemoryAccessToken.authenticate(accessToken.refreshToken))
            .to.be.null;

        user.isActive = false;
        // Event though saving is not required in memory storage backend it's
        // here to keep the code snippet a valid sample for all storage backends
        user.save();
        expect(await MemoryAccessToken.authenticate(accessToken.token))
            .to.be.null;
        user.isActive = true;
        // Event though saving is not required in memory storage backend it's
        // here to keep the code snippet a valid sample for all storage backends
        user.save();
        expect(await MemoryAccessToken.authenticate(accessToken.token))
            .to.be.equal(user);
    });
});

describe('Refreshing token', (): void => {
    it('should be able to create a user, issue an access token for her and' +
        ' refresh that token (consuming old one). Refreshing token should' +
        ' return null called with invalid token', async (): Promise<void> => {
        const user = new MemoryUser({username: 'u', password: 'p'});
        await user.save();

        const accessToken = await MemoryAccessToken.issue(user);
        expect(await MemoryAccessToken.refreshToken('some-token'))
            .to.be.null;
        expect(await MemoryAccessToken.refreshToken(accessToken.token))
            .to.be.null;

        expect(await MemoryAccessToken.refreshToken(accessToken.refreshToken))
            .to.be.null;
        user.isActive = true;
        // Event though saving is not required in memory storage backend it's
        // here to keep the code snippet a valid sample for all storage backends
        user.save();
        const result = await MemoryAccessToken.refreshToken(
            accessToken.refreshToken,
        );
        expect(result)
            .to.be.an.instanceOf(MemoryAccessToken);
        expect(accessToken.isConsumed())
            .to.be.true;
    });
});


src/models/tests/user.test.ts
import chai from 'chai';
import chaiAsPromised from 'chai-as-promised';
import sinon from 'sinon';
import sinonChai from 'sinon-chai';
import crypto from 'crypto';

import User from '../user';
import {AccessToken} from '..';

chai.use(chaiAsPromised);
chai.use(sinonChai);
const expect = chai.expect;

afterEach((): void => sinon.restore());

class DummyUser extends User {
    public id = 1;
    public username: string;
    public password: string;
    public isActive: boolean;
    public createdAt = new Date();
    public updatedAt = new Date();

    public constructor({username, password}: {
        username: string;
        password: string;
    }) {
        super();
        this.username = username;
        this.password = User.hashPassword(password);
        this.isActive = false; // TODO: Configuration
    }

    public save(): Promise<void> {
        return new Promise((): void => {});
    }

    public remove(): Promise<void> {
        return new Promise((): void => {});
    }

    public async getAccessToken(token: string): Promise<AccessToken|null> {
        return new Promise((): void => {});
    }

    public async getAccessTokens(): Promise<AccessToken[]> {
        return new Promise((): void => {});
    }

    public async getActiveAccessTokens(): Promise<AccessToken[]> {
        return new Promise((): void => {});
    }
}

describe('User class', (): void => {
    it('should exist', (): void => {
        expect(User)
            .to.exist;
    });

    it('should define but not implement "first" static method' +
        '', async (): Promise<void> => {
        expect(User.first)
            .to.be.a('function');
        await expect(User.first({id: 1}))
            .to.be.rejectedWith('"first" method is not implemented for "User"');
    });

    describe('hashPassword method', (): void => {
        it('hashes passwords with random salt when called with single' +
            ' argument', (): void => {
            expect(User.hashPassword)
                .to.be.a('function');

            const password = 'password';
            const algorithm = 'sha256';
            const iterations = 240000;
            const keyLength = 32;

            const result = User.hashPassword(password);

            const salt = result.split('$')[2];
            const hash = crypto.pbkdf2Sync(
                password,
                salt,
                iterations,
                keyLength,
                algorithm,
            ).toString('base64');
            const expectedHashedPassword =
                `${algorithm}$${iterations}$${salt}$${hash}`;
            expect(result)
                .to.be.equal(expectedHashedPassword);
        });

        it('hashes passwords with given salt when called with two arguments' +
            '', (): void => {
            expect(User.hashPassword)
                .to.be.a('function');

            const password = 'password';
            const algorithm = 'sha256';
            const iterations = 240000;
            const keyLength = 32;
            const salt = crypto.randomBytes(9).toString('base64');

            const result = User.hashPassword(password, salt);

            const hash = crypto.pbkdf2Sync(
                password,
                salt,
                iterations,
                keyLength,
                algorithm,
            ).toString('base64');
            const expectedHashedPassword =
                `${algorithm}$${iterations}$${salt}$${hash}`;
            expect(result)
                .to.be.equal(expectedHashedPassword);
        });
    });

    it('should implement a "setPassword" method that sets the salted hashed' +
        ' password as password field of the instance' +
        '', async (): Promise<void> => {
        const username = 'username';
        const password = 'password';
        const user = new DummyUser({username, password});
        sinon.replace(User, 'hashPassword', sinon.fake.returns('hashed'));
        sinon.replace(user, 'save', sinon.fake());

        await user.setPassword(password);

        expect(User.hashPassword)
            .to.have.been.calledOnceWith(password);
        expect(user.save)
            .to.have.been.calledOnce;
        expect(user.password)
            .to.be.equal('hashed');
    });

    it('should implement a "checkPassword" method that takes a password and' +
        ' returns true iff it matches the user\'s password', (): void => {
        const username = 'username';
        const password = 'password';
        const user = new DummyUser({username, password});

        sinon.spy(User, 'hashPassword');

        const result = user.checkPassword(password);

        expect(User.hashPassword)
            .to.have.been.calledOnceWith(password);
        expect(result)
            .to.be.ok;
    });

    it('should implement an "authenticate" method that takes a username and a' +
        ' password and returns the matching user if credentials are valid' +
        '', async (): Promise<void> => {
        const username = 'username';
        const password = 'password';
        const user = new DummyUser({username, password});
        user.isActive = true;

        sinon.replace(User, 'first', sinon.fake(
            ({username}: {username: string}): User|null => {
                if (username === user.username)
                    return user;
                return null;
            },
        ));
        sinon.spy(User, 'hashPassword');

        const result = await User.authenticate(username, password);

        expect(User.first)
            .to.have.been.calledOnceWith({username, isActive: true});
        expect(User.hashPassword)
            .to.have.been.calledOnceWith(password);
        expect(result)
            .to.be.equal(user);
    });

    it('should implement an "authenticate" method that takes a username and a' +
        ' password and returns null if invalid username is provided' +
        '', async (): Promise<void> => {
        const username = 'username';
        const password = 'password';
        const user = new DummyUser({username, password});
        user.isActive = true;

        sinon.replace(User, 'first', sinon.fake(
            ({username}: {username: string}): User|null => {
                if (username === user.username)
                    return user;
                return null;
            },
        ));

        const result = await User.authenticate('invalid-username', password);

        expect(User.first)
            .to.have.been.calledOnceWith({
                username: 'invalid-username',
                isActive: true,
            });
        expect(result)
            .to.be.null;
    });

    it('should implement an "authenticate" method that takes a username and a' +
        ' password and returns null if invalid password is provided' +
        '', async (): Promise<void> => {
        const username = 'username';
        const password = 'password';
        const user = new DummyUser({username, password});
        user.isActive = true;

        sinon.replace(User, 'first', sinon.fake(
            ({username}: {username: string}): User|null => {
                if (username === user.username)
                    return user;
                return null;
            },
        ));

        const result = await User.authenticate(username, 'invalid-password');

        expect(User.first)
            .to.have.been.calledOnceWith({username, isActive: true});
        expect(result)
            .to.be.null;
    });

    it('should implement an "authenticate" method that takes a username and a' +
        ' password and returns null if the user is inactivated' +
        '', async (): Promise<void> => {
        const username = 'username';
        const password = 'password';
        const user = new DummyUser({username, password});

        sinon.replace(User, 'first', sinon.fake(
            ({username}: {username: string}): User|null => {
                if (username === user.username)
                    return user;
                return null;
            },
        ));

        const result = await User.authenticate(username, password);

        expect(User.first)
            .to.have.been.calledOnceWith({username, isActive: true});
        expect(result)
            .to.be.null;
    });

    it('should implement a "logout" method that takes an access token string' +
        ' and in case it\'s an active access token of the user, it should' +
        ' revoke it', async (): Promise<void> => {
        const token = 'some-token';
        const user = new DummyUser({
            username: 'username',
            password: 'password',
        });
        const accessToken: AccessToken = {
            token,
            revoke: sinon.fake(),
            isActive: (): boolean => true,
        } as unknown as AccessToken;
        sinon.replace(user, 'getAccessToken', sinon.fake(
            (token: string): AccessToken|null => {
                if (token === accessToken.token)
                    return accessToken;
                return null;
            },
        ));

        await user.logout(token);

        expect(user.getAccessToken)
            .to.have.been.calledOnceWith(token);
        expect(accessToken.revoke)
            .to.have.been.calledOnce;
    });

    it('should implement a "globalLogout" method that revokes all active' +
        ' access tokens of the user', async (): Promise<void> => {
        const user = new DummyUser({
            username: 'username',
            password: 'password',
        });
        const accessTokens: AccessToken[] = [
            {token: '0', revoke: sinon.fake(), user},
            {token: '1', revoke: sinon.fake(), user},
            {token: '2', revoke: sinon.fake(), user},
        ] as unknown as AccessToken[];
        sinon.replace(user, 'getActiveAccessTokens', sinon.fake(
            (): AccessToken[] => {
                return accessTokens;
            },
        ));

        await user.globalLogout();

        expect(user.getActiveAccessTokens)
            .to.have.been.calledOnce;
        for (const accessToken of accessTokens)
            expect(accessToken.revoke)
                .to.have.been.calledOnce;
    });
});


src/memory-storage/tests/user.test.ts
import chai from 'chai';
import chaiAsPromised from 'chai-as-promised';
import sinon from 'sinon';
import sinonChai from 'sinon-chai';
import faker from 'faker';
import each from 'jest-each';

import {MemoryUser} from '..';
import {AccessToken} from '../..';

chai.use(chaiAsPromised);
chai.use(sinonChai);
const expect = chai.expect;

afterEach((): void => {
    MemoryUser.items = {};
    sinon.restore();
});

describe('StoredUser class', (): void => {
    it('should exist', (): void => {
        expect(MemoryUser)
            .to.exist;
    });

    it('should be constructable', (): void => {
        sinon.replace(MemoryUser, 'hashPassword', <T>(a: T): T => a);
        const username = faker.internet.userName();
        const password = faker.internet.password();
        const user = new MemoryUser({username, password});

        expect(user)
            .to.have.property('username')
            .that.is.a('string')
            .that.is.equal(username);
        expect(user)
            .to.have.property('password')
            .that.is.a('string')
            .that.satisfies((password: string): boolean =>
                (password.match(/\$/g) || []).length === 3);
    });

    it('should be able to save instances which includes generating an id for'
        + ' them', (): void => {
        const username = faker.internet.userName();
        const password = faker.internet.password();
        const user = new MemoryUser({username, password});
        user.save();

        expect(user)
            .to.have.property('id')
            .that.is.a('number');
    });

    it('should be able to remove instances which includes unsetting their id'
        + ' attribute', (): void => {
        const username = faker.internet.userName();
        const password = faker.internet.password();
        const user = new MemoryUser({username, password});
        user.save();

        expect(user)
            .to.have.property('id')
            .that.is.a('number');

        user.remove();

        expect(user)
            .to.have.property('id')
            .that.is.undefined;
    });

    it('should implement "getAccessToken" method', (): void => {
        const username = 'username';
        const password = 'password';
        const token = 'token';
        const user = new MemoryUser({username, password});
        expect(user.getAccessToken)
            .to.be.a('function');
        sinon.replace(AccessToken, 'first', sinon.fake());

        user.getAccessToken(token);

        expect(AccessToken.first)
            .to.be.calledOnceWith({token, userId: user.id});
    });

    it('should implement "getAccessTokens" method', (): void => {
        const username = 'username';
        const password = 'password';
        const user = new MemoryUser({username, password});
        expect(user.getAccessTokens)
            .to.be.a('function');
        sinon.replace(AccessToken, 'filter', sinon.fake());

        user.getAccessTokens();

        expect(AccessToken.filter)
            .to.be.calledOnceWith({userId: user.id});
    });

    it('should implement "getAccessToken" method', (): void => {
        const username = 'username';
        const password = 'password';
        const user = new MemoryUser({username, password});
        expect(user.getActiveAccessTokens)
            .to.be.a('function');
        sinon.replace(AccessToken, 'filter', sinon.fake());

        user.getActiveAccessTokens();

        expect(AccessToken.filter)
            .to.be.calledOnceWith({active: true, userId: user.id});
    });

    describe('should implement a "first" method that looks up a "User"' +
        ' in database based on its token and optionally user id or expiration' +
        ' status.', (): void => {
        const users = [
            new MemoryUser({username: 'user1', password: 'password'}),
            new MemoryUser({username: 'user2', password: 'password'}),
            new MemoryUser({username: 'user3', password: 'password'}),
        ];
        each([
            [users[0], {username: users[0].username}, users],
            [users[1], {username: users[1].username}, users],
            [users[2], {id: (): number => users[2].id}, users],
            [
                null,
                {id: (): number => users[0].id + users[1].id + users[2].id},
                users,
            ],
            [null, {username: 'invalid-username'}, users],
            [users[0], {}, users],
            [
                null,
                {username: users[0].username, id: (): number => users[1].id},
                users,
            ],
        ]).it('should return %s for query %s', async (
            expected,
            query,
            users,
        ): Promise<void> => {
            if (query.id instanceof Function)
                query.id = query.id();
            await Promise.all(users.map(
                async (user: MemoryUser): Promise<void> => await user.save(),
            ));
            const result = await MemoryUser.first(query);

            expect(result)
                .to.be.equal(expected);
        });
    });
});


src/models/access-token.ts
import {User} from './index';

interface Constructor<T = {}> {
    new(...args: any[]): T;
}

/**
 * Issuing and authenticating tokens is the main purpose of this class.
 * To be subclassed by a class that implements storage methods (save, get and
 * remove)
 */
export default abstract class AccessToken {
    /**
     * Primary key
     */
    public abstract id: number;
    /**
     * Token string - autogenerated no need to be filled
     */
    public abstract token: string;
    /**
     * Refresh token string - autogenerated no need to be filled
     */
    public abstract refreshToken: string;
    /**
     * Timestamp in which this token will get expired, it only affects `token`
     * and has no effect on `refreshToken`
     */
    public abstract expires: Date|null;
    /**
     * The user that this access token belongs to
     */
    public abstract user: User;
    /**
     * Indicates that whether `refreshToken` is used or not, as soon as it's
     * used this field will be true and the `AccessToken` instance will become
     * inactive
     */
    protected abstract consumed: boolean;
    /**
     * Timestamp of the creation of the instance
     */
    public abstract createdAt: Date;
    /**
     * Timestamp of the last update on the instance (last time save method has
     * been called)
     */
    public abstract updatedAt: Date;

    /**
     * Creates an AccessToken for a user setting its expiration date
     */
    public constructor(...args: any[]) {
    }

    /**
     * Saves the AccessToken instance in databse
     */
    public abstract async save(): Promise<void>;

    /**
     * Looks an AccessToken up in database based on its token, refreshToken,
     * userId, its expired status, its consumed status or its active status
     */
    public static async first(filters: {
        /** Token */
        token?: string;
        /** Refresh token */
        refreshToken?: string;
        /** User */
        userId?: number;
        /** Expired */
        expired?: boolean;
        /** Consumed */
        consumed?: boolean;
        /** Active */
        active?: boolean;
    }): Promise<AccessToken|null> {
        throw new Error(`"first" method is not implemented for "${this.name}"`);
    }

    /**
     * Looks up AccessTokens in database based on its userId,
     * its expired status, its consumed status or its active status
     *
     * @param {Object} filters Filters object
     * @param {number} filters.userId User
     * @param {boolean} filters.expired Expired
     * @param {boolean} filters.consumed Consumed
     * @param {boolean} filters.active Active
     */
    public static async filter(filters: {
        /** User */
        userId?: number;
        /** Expired */
        expired?: boolean;
        /** Consumed */
        consumed?: boolean;
        /** Active */
        active?: boolean;
    }): Promise<AccessToken[]> {
        throw new Error(`"filter" method is not implemented for "${this.name}"`);
    }

    /**
     * Removes the AccessToken from database
     */
    public abstract remove(): Promise<void>;

    /**
     * Checks if this AccessToken is active (not expired and not consumed)
     */
    public isActive(): boolean {
        return !this.isExpired() && !this.isConsumed();
    }

    /**
     * Checks if this AccessToken is expired, returns true if expired and
     * returns false otherwise.
     */
    public isExpired(): boolean {
        return this.expires === null ? true : this.expires <= new Date();
    }

    /**
     * Checks if this AccessToken is consumed, returns true if consumed and
     * returns false otherwise.
     */
    public isConsumed(): boolean {
        return this.consumed;
    }

    /**
     * Revokes this AccessToken
     */
    public async revoke(): Promise<void> {
        return await this.remove();
    }

    /**
     * Issue a new pair of access token and refresh token
     */
    public static async issue(user: User): Promise<AccessToken> {
        // TODO: Configurable expiration time
        const constructor = this as (typeof AccessToken & Constructor);

        const accessToken = new constructor({
            user,
            expires: new Date(new Date().getTime() + 10 * 60 * 1000),
        }) as AccessToken;
        await accessToken.save();
        return accessToken;
    }

    /**
     * Authenticates a token. If successful returns a user
     * otherwise returns null
     */
    public static async authenticate(token: string): Promise<User|null> {
        const accessToken = await this.first({token});
        if (
            accessToken && accessToken.isActive() &&
            accessToken.user && accessToken.user.isActive
        ) {
            return accessToken.user;
        }
        return null;
    }

    /**
     * If an active "AccessToken" with its "refreshToken" equal to the token
     * parameter of this method, it'll get consumed, a new "AccessToken" will
     * be issued and returns. Otherwise it's a no-op and returns null.
     */
    public static async refreshToken(
        refreshToken: string,
    ): Promise<AccessToken|null> {
        const accessToken = await this.first({refreshToken});
        if (
            accessToken && accessToken.isActive() &&
            accessToken.user && accessToken.user.isActive
        ) {
            accessToken.consumed = true;
            await accessToken.save();
            return await this.issue(accessToken.user);
        }
        return null;
    }
}


src/models/refresh-token.ts
export default abstract class RefreshToken {
    /**
     * save
     *
     * Saves the RefreshToken instance in databse
     */
    public abstract save(): Promise<void>;

    /**
     * first
     *
     * Looks an RefreshToken up in database based on its token, accessTokenId,
     * userId or consumed attributes
     */
    public static first({token, accessTokenId, userId, consumed}: {
        token?: string;
        accessTokenId?: number;
        userId?: number;
        consumed?: boolean;
    }): Promise<RefreshToken|null> {
        throw new Error('Not implemented');
    }

    /**
     * filter
     *
     * Looks up RefreshTokens in database based on their userId or consumed
     * attributes
     */
    public static filter({userId, consumed}: {
        userId?: number;
        consumed?: boolean;
    }): Promise<RefreshToken[]> {
        throw new Error('Not implemented');
    }

    /**
     * remove
     *
     * Removes the RefreshToken from database
     */
    public abstract remove(): Promise<void>;
}


src/models/user.ts
import crypto from 'crypto';
import {AccessToken} from './index';

/**
 * User abstract class to be subclassed by a class that implements storage
 * methods (save, get and remove)
 */
export default abstract class User {
    /**
     * Primary key
     */
    public abstract id: number;
    /**
     * User's username (unique)
     */
    public abstract username: string;
    /**
     * User's password stored as a hash
     */
    public abstract password: string;
    /**
     * Stores whether the user is active or not in a boolean
     */
    public abstract isActive: boolean;
    /**
     * Timestamp of the creation of the instance
     */
    public abstract createdAt: Date;
    /**
     * Timestamp of the last update on the instance (last time save method has
     * been called)
     */
    public abstract updatedAt: Date;

    /**
     * Creates a User by its username and password
     */
    public constructor(...args: any[]) {
    }

    /**
     * Saves the User instance in databse
     */
    public abstract async save(): Promise<void>;

    /**
     * Looks a User up in database based on its id or username
     *
     * @param {Object} filters Filters object
     * @param {number} filters.id Id
     * @param {strign} filters.username Username
     * @param {boolean} filters.isActive Activation status
     */
    public static async first(filters: {
        id?: number;
        username?: string;
        isActive?: boolean;
    }): Promise<User|null> {
        throw new Error(`"first" method is not implemented for "${this.name}"`);
    }

    /**
     * Removes the User instance from database
     */
    public abstract async remove(): Promise<void>;

    /**
     * Takes a string token and returns the AccessToken instance with that
     * token that belongs to the User instance (has its user foreign key set to
     * the User instance).
     */
    public abstract async getAccessToken(token: string):
    Promise<AccessToken|null>;

    /**
     * Returns all AccessToken instances that belong to this User (have their
     * user foreign key set to the User instance).
     */
    public abstract async getAccessTokens(): Promise<AccessToken[]>;

    /**
     * Returns all AccessToken instances that belong to this User (have their
     * user foreign key set to the User instance) and are not expired yet.
     */
    public abstract async getActiveAccessTokens(): Promise<AccessToken[]>;

    /**
     * Takes a password and returns a salted hash of it
     */
    public static hashPassword(
        password: string,
        salt?: string,
    ): string { // TODO: config
        const algorithm = 'sha256';
        const iterations = 240000;
        const keyLength = 32;
        // In order to achieve mostly unique salt per user, we generate 72 bit
        // salts here.
        if (!salt)
            salt = crypto.randomBytes(9).toString('base64');
        const hash = crypto.pbkdf2Sync(
            password,
            salt,
            iterations,
            keyLength,
            algorithm,
        ).toString('base64');
        return `${algorithm}$${iterations}$${salt}$${hash}`;
    }

    /**
     * Set the User's password and saves the User instance. It stores a salted
     * hashed version of password instead, raw password is never saved.
     */
    public async setPassword(password: string): Promise<void> {
        this.password = User.hashPassword(password);
        return await this.save();
    }

    /**
     * As raw passwords are not saved, this method is useful to check a raw
     * password against the hashed password of the User stored in database.
     * If passwords match it reutns true, otherwise it returns false.
     */
    public checkPassword(password: string): boolean {
        return this.password === User.hashPassword(password, this.password.split('$')[2]);
    }

    /**
     * Takes a username and password and checks if they match with a User.
     * If it matches it returns the matching User, otherwise it returns null
     */
    public static async authenticate(
        username: string,
        password: string,
    ): Promise<User|null> {
        const user = await this.first({username, isActive: true});
        // The value of "isActive" is double checked in case the class that
        // implements "first" method mistakenly is not respecting "isActive"
        // parameter. It's preferred that it respects "isActive" for
        // performance concerns but in case it doesn't we double check here.
        if (!user || !user.isActive) {
            // This is required to keep the response time of this function the
            // same whether the user exists or not. So that an attacker can't
            // understand if the user exists or not by measuring the response
            // time
            User.hashPassword(password);
            return null;
        }
        if (!user.checkPassword(password)) {
            return null;
        }
        return user;
    }

    /**
     * Takes a token string. If there's an active access token associated with
     * this User with the provided token, it'll revoke it, otherwise it'll be a
     * no-op.
     */
    public async logout(token: string): Promise<void> {
        const accessToken = await this.getAccessToken(token);
        if (accessToken && accessToken.isActive()) {
            await accessToken.revoke();
        }
    }

    /**
     * Revokes all active access tokens of the User so that she's logged out of
     * all devices she's currently logged in.
     */
    public async globalLogout(): Promise<void> {
        const accessTokens = await this.getActiveAccessTokens();
        for (const accessToken of accessTokens) {
            await accessToken.revoke();
        }
    }
}
